network/serde/jsonrpc.go

```diff
@@ -17,60 +31,49 @@ const (
 	TypeError    MessageType = 3
 )
 
-type JsonRpcRequest struct {
-	Jsonrpc string      `json:"jsonrpc"`
-	Id      uint64      `json:"id"`
-	Method  string      `json:"method"`
-	Params  interface{} `json:"params"`
+type JsonRpcRequestResponse struct {
+	Jsonrpc string        `json:"jsonrpc"`
+	Id      uint64        `json:"id"`
+	Method  RequestMethod `json:"method"`
+	Params  interface{}   `json:"params"`
+	Result  interface{}   `json:"result"`
+	Error   interface{}   `json:"error"`
 }
-
-type JsonRpcResponse struct {
-	Jsonrpc string      `json:"jsonrpc"`
-	Id      uint64      `json:"id"`
-	Result  interface{} `json:"result"`
-	Error   interface{} `json:"error"`
+type RequestPayload interface {
+	directfund.ObjectiveRequest | directdefund.ObjectiveRequest | virtualfund.ObjectiveRequest | virtualdefund.ObjectiveRequest
 }
-
-type JsonRpcRequestResponse struct {
+type JsonRpcRequest[T RequestPayload] struct {
+	Jsonrpc string `json:"jsonrpc"`
+	Id      uint64 `json:"id"`
+	Method  string `json:"method"`
+	Params  T      `json:"params"`
+}
+type ResponsePayload interface {
+	directfund.ObjectiveResponse | protocols.ObjectiveId | virtualfund.ObjectiveResponse
+}
+type JsonRpcResponse[T ResponsePayload] struct {
 	Jsonrpc string      `json:"jsonrpc"`
 	Id      uint64      `json:"id"`
-	Method  string      `json:"method"`
-	Params  interface{} `json:"params"`
-	Result  interface{} `json:"result"`
+	Result  T           `json:"result"`
 	Error   interface{} `json:"error"`
 }
 
-type JsonRpcDirectFundRequest struct {
-	Jsonrpc          string                      `json:"jsonrpc"`
-	Id               uint64                      `json:"id"`
-	Method           string                      `json:"method"`
-	ObjectiveRequest directfund.ObjectiveRequest `json:"params"`
-}
-
-type JsonRpcDirectFundResponse struct {
-	Jsonrpc           string                       `json:"jsonrpc"`
-	Id                uint64                       `json:"id"`
-	ObjectiveResponse directfund.ObjectiveResponse `json:"result"`
-	Error             interface{}                  `json:"error"`
-}
-
-type JsonRpc struct{}
+func NewJsonRpcRequest[T RequestPayload](requestId uint64, method RequestMethod, objectiveRequest T) *JsonRpcRequest[T] {
```

Comment:
Ideally we'd just infer `method` from the type of `objectiveRequest`. However it seems in golang you can't switch on the type of a generic parameter 😞